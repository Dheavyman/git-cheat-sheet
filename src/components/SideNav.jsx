import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';

import AuthForm from './AuthForm';
import CommentForm from './CommentForm';
import { authenticate, logout } from '../actions/users';

/**
 * SideNav component
 *
 * @className SideNav
 * @extends {Component}
 */
export class SideNav extends Component {
  constructor(props) {
    super(props);
    this.state = {
      forLogin: true,
      username: '',
      password: '',
      showForm: false,
    }
    this.initializeSidenav = this.initializeSidenav.bind(this);
    this.handleToggleAuth = this.handleToggleAuth.bind(this);
    this.handleShowForm = this.handleShowForm.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleLogout = this.handleLogout.bind(this);
  }

  componentDidMount() {
    this.initializeSidenav()
  }

  componentDidUpdate() {
    this.initializeSidenav();
  }

  initializeSidenav() {
    document.addEventListener('DOMContentLoaded', () => {
      // eslint-disable-next-line no-undef
      M.Sidenav.init(document.querySelectorAll('.sidenav'), {
        edge: 'right',
      });
    });
  }

  /**
   * Handle display of login or register button
   *
   * @memberof SideNav
   */
  handleToggleAuth() {
    this.setState({
      forLogin: !this.state.forLogin
    })
  }

  /**
   * Handle display of form component
   *
   * @memberof SideNav
   */
  handleShowForm() {
    this.setState({
      showForm: !this.state.showForm
    });
  }

  /**
   * Handle form input change
   *
   * @param {object} event - Event object
   * @memberof SideNav
   */
  handleChange(event) {
    const { target: { name, value } } = event;
    this.setState({
      [name]: value
    })
  }

  /**
   * Handle submitting login or register event
   *
   * @param {object} event - event object
   * @memberof SideNav
   */
  handleSubmit(event) {
    event.preventDefault();
    const { forLogin, username, password } = this.state;
    const { authenticate } = this.props;

    if (forLogin) {
      authenticate({ username, password }, 'Login')
        .then(() => {
          if (this.props.isLoggedIn) {
            this.setState({
              showForm: !this.state.showForm
            })
          }
        })
    } else {
      authenticate({ username, password }, 'Register')
        .then(() => {
          if (this.props.isLoggedIn) {
            this.setState({
              showForm: !this.state.showForm
            })
          }
        })
    }
  }

  handleLogout() {
    this.props.logout();
  }

  render() {
    const { forLogin, showForm } = this.state;
    const { isLoading, isLoggedIn, user } = this.props;

    return (
      <Fragment>
        <ul id="slide-out" className="sidenav">
          <li className="git-logo">
            {isLoggedIn &&
            <a
              href="#!"
              className="logout"
              onClick={this.handleLogout}
            >
              Logout
            </a>}
            <img
              src="https://res.cloudinary.com/heavyman/image/upload/v1544835375/Git-Icon-Black_b2dvxp.png"
              alt="" className="logo-img"
            />
            <p>The awesome git cheat sheet</p>
          </li>
          {isLoggedIn && (
            <li className="user">
              <p className="username">
                Hi <span>{user && user.username}</span>
              </p>
            </li>
          )}
          <li><div className="divider"></div></li>
          {showForm
          ? <AuthForm
              forLogin={forLogin}
              isLoading={isLoading}
              handleToggleAuth={this.handleToggleAuth}
              handleChange={this.handleChange}
              handleSubmit={this.handleSubmit}
            />
          : <li className="info-text">
            <p>Whether you are new to git or just needing a refresher,
              this cheat sheet will help you discover or remember useful
              git commands.
            </p>
            {!isLoggedIn && (
            <p className="center-align">
              Click <a
                href="#!"
                className="cursor-pointer blue-text"
                onClick={this.handleShowForm}
                >here </a>
                for more privilege.
            </p>
            )}
          </li>
          }
          {isLoggedIn && <CommentForm />}
        </ul>
      </Fragment>
    );
  }
};

export default connect(({ userReducer: { isLoading, isLoggedIn, user } }) => ({
  isLoading: isLoading,
  isLoggedIn: isLoggedIn,
  user: user
}), {authenticate, logout})(SideNav);
