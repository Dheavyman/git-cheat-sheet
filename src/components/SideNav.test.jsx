import React from 'react';
import TestRenderer from 'react-test-renderer';

import { SideNav } from './SideNav';

describe('SideNav component', () => {
  let props;

  beforeEach(() => {
    props = {
      isLoading: false,
      isLoggedIn: false,
      user: {
        username: 'username',
      },
      authenticate: jest.fn(() => Promise.resolve()),
      logout: jest.fn()
    }
  });

  it('should render component successfully', () => {
    const renderer = TestRenderer.create(<SideNav {...props} />);

    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('should show login form when user clicks to login', () => {
    const renderer = TestRenderer.create(<SideNav {...props} />);
    const handleShowFormSpy = jest.spyOn(renderer.getInstance(),
    'handleShowForm');

    renderer.update(<SideNav />);

    expect(renderer.root.instance.state.showForm).toBe(false);

    renderer.root.findByProps({className:"cursor-pointer blue-text"})
      .props.onClick();
    expect(handleShowFormSpy).toBeCalled();
    expect(renderer.root.instance.state.showForm).toBe(true);
    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('should call handle change when input value changes', () => {
    const renderer = TestRenderer.create(<SideNav {...props} />);
    const handleChangeSpy = jest.spyOn(renderer.getInstance(), 'handleChange');

    renderer.update(<SideNav />);

    expect(renderer.root.instance.state.username).toBe('');
    expect(renderer.root.instance.state.password).toBe('');

    renderer.root.findByProps({className:"cursor-pointer blue-text"})
      .props.onClick();
    renderer.root.findByProps({name:"username"}).props.onChange(
      { target: { name: 'username', value: 'Mathew'}});
    renderer.root.findByProps({name:"password"}).props.onChange(
      { target: { name: 'password', value: 'secret'}});

    expect(handleChangeSpy).toBeCalled();
    expect(renderer.root.instance.state.username).toBe('Mathew');
    expect(renderer.root.instance.state.password).toBe('secret');
    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('should call handleToggleAuth to show login or register', () => {
    const renderer = TestRenderer.create(<SideNav {...props} />);
    const handleToggleAuthSpy = jest.spyOn(renderer.getInstance(),
    'handleToggleAuth');

    renderer.update(<SideNav />);

    expect(renderer.root.instance.state.forLogin).toBe(true);

    renderer.root.findByProps({className:"cursor-pointer blue-text"})
      .props.onClick();
    renderer.root.findByType('a').props.onClick();

    expect(handleToggleAuthSpy).toBeCalled();
    expect(renderer.root.instance.state.forLogin).toBe(false);
    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('should call handleSubmit for login event', () => {
    const renderer = TestRenderer.create(<SideNav {...props} />);
    const handleSubmitSpy = jest.spyOn(renderer.getInstance(), 'handleSubmit');

    renderer.update(<SideNav {...props} />);
    renderer.root.findByProps({className:"cursor-pointer blue-text"})
      .props.onClick();

    expect(renderer.root.findByType('button').props.type).toEqual('submit');
    expect(renderer.root.findByType('button').props.children).toEqual('Login');
    expect(renderer.root.instance.state.forLogin).toBe(true)

    renderer.root.instance.state.username = 'Anonymous';
    renderer.root.instance.state.password = 'password';
    renderer.root.findByType('form').props
      .onSubmit({ preventDefault: jest.fn() });

    props = {...props, isLoggedIn: true}
    renderer.update(<SideNav {...props} />)

    expect(handleSubmitSpy).toBeCalled();
    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('should call handleSubmit for register event', () => {
    const renderer = TestRenderer.create(<SideNav {...props} />);
    const handleSubmitSpy = jest.spyOn(renderer.getInstance(), 'handleSubmit');

    renderer.update(<SideNav {...props} />);
    renderer.root.findByProps({className:"cursor-pointer blue-text"})
      .props.onClick();
    renderer.root.findByType('a').props.onClick();

    expect(renderer.root.findByType('button').props.type).toEqual('submit');
    expect(renderer.root.findByType('button').props.children).toEqual('Register');
    expect(renderer.root.instance.state.forLogin).toBe(false)

    renderer.root.instance.state.username = 'Anonymous1';
    renderer.root.instance.state.password = 'password1';
    renderer.root.findByType('form').props
      .onSubmit({ preventDefault: jest.fn() });

      props = {...props, isLoggedIn: true}
      renderer.update(<SideNav {...props} />)

    expect(handleSubmitSpy).toBeCalled();
    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('should call handleLogout for logout event', () => {
    props = {
      ...props,
      isLoggedIn: true,
    };
    const renderer = TestRenderer.create(<SideNav {...props} />);
    const handleLogoutSpy = jest.spyOn(renderer.getInstance(), 'handleLogout');

    renderer.update(<SideNav {...props} />);

    renderer.root.findByProps({className:"logout"}).props.onClick();

    expect(handleLogoutSpy).toBeCalled();
    expect(renderer.toJSON()).toMatchSnapshot();
  });
});
