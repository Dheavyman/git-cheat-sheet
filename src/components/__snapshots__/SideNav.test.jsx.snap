// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNav component should call handle change when input value changes 1`] = `
<ul
  className="sidenav"
  id="slide-out"
>
  <li
    className="git-logo"
  >
    <img
      alt=""
      className="logo-img"
      src="https://res.cloudinary.com/heavyman/image/upload/v1544835375/Git-Icon-Black_b2dvxp.png"
    />
    <p>
      The awesome git cheat sheet
    </p>
  </li>
  <li>
    <div
      className="divider"
    />
  </li>
  <div
    className="row"
  >
    <form
      className="col s12"
      onSubmit={[Function]}
    >
      <div
        className="row"
      >
        <div
          className="input-field col s12"
        >
          <input
            className="validate"
            id="username"
            name="username"
            onChange={
              [MockFunction] {
                "calls": Array [
                  Array [
                    Object {
                      "target": Object {
                        "name": "username",
                        "value": "Mathew",
                      },
                    },
                  ],
                  Array [
                    Object {
                      "target": Object {
                        "name": "password",
                        "value": "secret",
                      },
                    },
                  ],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                ],
              }
            }
            required={true}
            type="text"
          />
          <label
            htmlFor="username"
          >
            Username
          </label>
        </div>
        <div
          className="input-field col s12"
        >
          <input
            className="validate"
            id="password"
            name="password"
            onChange={
              [MockFunction] {
                "calls": Array [
                  Array [
                    Object {
                      "target": Object {
                        "name": "username",
                        "value": "Mathew",
                      },
                    },
                  ],
                  Array [
                    Object {
                      "target": Object {
                        "name": "password",
                        "value": "secret",
                      },
                    },
                  ],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                ],
              }
            }
            required={true}
            type="password"
          />
          <label
            htmlFor="password"
          >
            Password
          </label>
        </div>
        
        <div
          className="row center-align"
        >
          <button
            className="waves-effect waves-light btn"
            type="submit"
          >
            Login
          </button>
        </div>
        <div
          className="row center-align"
        >
          Don't have an account? 
          <a
            className="cursor-pointer blue-text"
            href="#!"
            onClick={[Function]}
          >
            Register
          </a>
        </div>
      </div>
    </form>
  </div>
</ul>
`;

exports[`SideNav component should call handleLogout for logout event 1`] = `
<ul
  className="sidenav"
  id="slide-out"
>
  <li
    className="git-logo"
  >
    <a
      className="logout"
      href="#!"
      onClick={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        }
      }
    >
      Logout
    </a>
    <img
      alt=""
      className="logo-img"
      src="https://res.cloudinary.com/heavyman/image/upload/v1544835375/Git-Icon-Black_b2dvxp.png"
    />
    <p>
      The awesome git cheat sheet
    </p>
  </li>
  <li
    className="user"
  >
    <p
      className="username"
    >
      Hi 
      <span>
        username
      </span>
    </p>
  </li>
  <li>
    <div
      className="divider"
    />
  </li>
  <li
    className="info-text"
  >
    <p>
      Whether you are new to git or just needing a refresher, this cheat sheet will help you discover or remember useful git commands.
    </p>
  </li>
  <div
    className="row"
  >
    <div
      className="input-field col s12"
    >
      <textarea
        className="materialize-textarea"
        disabled={true}
        id="comment"
      />
      <label
        htmlFor="comment"
      >
        Add comment
      </label>
    </div>
  </div>
</ul>
`;

exports[`SideNav component should call handleSubmit for login event 1`] = `
<ul
  className="sidenav"
  id="slide-out"
>
  <li
    className="git-logo"
  >
    <a
      className="logout"
      href="#!"
      onClick={[Function]}
    >
      Logout
    </a>
    <img
      alt=""
      className="logo-img"
      src="https://res.cloudinary.com/heavyman/image/upload/v1544835375/Git-Icon-Black_b2dvxp.png"
    />
    <p>
      The awesome git cheat sheet
    </p>
  </li>
  <li
    className="user"
  >
    <p
      className="username"
    >
      Hi 
      <span>
        username
      </span>
    </p>
  </li>
  <li>
    <div
      className="divider"
    />
  </li>
  <div
    className="row"
  >
    <form
      className="col s12"
      onSubmit={
        [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "preventDefault": [MockFunction] {
                  "calls": Array [
                    Array [],
                  ],
                  "results": Array [
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                  ],
                },
              },
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        }
      }
    >
      <div
        className="row"
      >
        <div
          className="input-field col s12"
        >
          <input
            className="validate"
            id="username"
            name="username"
            onChange={[Function]}
            required={true}
            type="text"
          />
          <label
            htmlFor="username"
          >
            Username
          </label>
        </div>
        <div
          className="input-field col s12"
        >
          <input
            className="validate"
            id="password"
            name="password"
            onChange={[Function]}
            required={true}
            type="password"
          />
          <label
            htmlFor="password"
          >
            Password
          </label>
        </div>
        
        <div
          className="row center-align"
        >
          <button
            className="waves-effect waves-light btn"
            disabled={false}
            type="submit"
          >
            Login
          </button>
        </div>
        <div
          className="row center-align"
        >
          Don't have an account? 
          <a
            className="cursor-pointer blue-text"
            href="#!"
            onClick={[Function]}
          >
            Register
          </a>
        </div>
      </div>
    </form>
  </div>
  <div
    className="row"
  >
    <div
      className="input-field col s12"
    >
      <textarea
        className="materialize-textarea"
        disabled={true}
        id="comment"
      />
      <label
        htmlFor="comment"
      >
        Add comment
      </label>
    </div>
  </div>
</ul>
`;

exports[`SideNav component should call handleSubmit for register event 1`] = `
<ul
  className="sidenav"
  id="slide-out"
>
  <li
    className="git-logo"
  >
    <a
      className="logout"
      href="#!"
      onClick={[Function]}
    >
      Logout
    </a>
    <img
      alt=""
      className="logo-img"
      src="https://res.cloudinary.com/heavyman/image/upload/v1544835375/Git-Icon-Black_b2dvxp.png"
    />
    <p>
      The awesome git cheat sheet
    </p>
  </li>
  <li
    className="user"
  >
    <p
      className="username"
    >
      Hi 
      <span>
        username
      </span>
    </p>
  </li>
  <li>
    <div
      className="divider"
    />
  </li>
  <div
    className="row"
  >
    <form
      className="col s12"
      onSubmit={
        [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "preventDefault": [MockFunction] {
                  "calls": Array [
                    Array [],
                  ],
                  "results": Array [
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                  ],
                },
              },
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        }
      }
    >
      <div
        className="row"
      >
        <div
          className="input-field col s12"
        >
          <input
            className="validate"
            id="username"
            name="username"
            onChange={[Function]}
            required={true}
            type="text"
          />
          <label
            htmlFor="username"
          >
            Username
          </label>
        </div>
        <div
          className="input-field col s12"
        >
          <input
            className="validate"
            id="password"
            name="password"
            onChange={[Function]}
            required={true}
            type="password"
          />
          <label
            htmlFor="password"
          >
            Password
          </label>
        </div>
        
        <div
          className="row center-align"
        >
          <button
            className="waves-effect waves-light btn"
            disabled={false}
            type="submit"
          >
            Register
          </button>
        </div>
        <div
          className="row center-align"
        >
          Already have an account? 
          <a
            className="cursor-pointer blue-text"
            href="#!"
            onClick={[Function]}
          >
            Login
          </a>
        </div>
      </div>
    </form>
  </div>
  <div
    className="row"
  >
    <div
      className="input-field col s12"
    >
      <textarea
        className="materialize-textarea"
        disabled={true}
        id="comment"
      />
      <label
        htmlFor="comment"
      >
        Add comment
      </label>
    </div>
  </div>
</ul>
`;

exports[`SideNav component should call handleToggleAuth to show login or register 1`] = `
<ul
  className="sidenav"
  id="slide-out"
>
  <li
    className="git-logo"
  >
    <img
      alt=""
      className="logo-img"
      src="https://res.cloudinary.com/heavyman/image/upload/v1544835375/Git-Icon-Black_b2dvxp.png"
    />
    <p>
      The awesome git cheat sheet
    </p>
  </li>
  <li>
    <div
      className="divider"
    />
  </li>
  <div
    className="row"
  >
    <form
      className="col s12"
      onSubmit={[Function]}
    >
      <div
        className="row"
      >
        <div
          className="input-field col s12"
        >
          <input
            className="validate"
            id="username"
            name="username"
            onChange={[Function]}
            required={true}
            type="text"
          />
          <label
            htmlFor="username"
          >
            Username
          </label>
        </div>
        <div
          className="input-field col s12"
        >
          <input
            className="validate"
            id="password"
            name="password"
            onChange={[Function]}
            required={true}
            type="password"
          />
          <label
            htmlFor="password"
          >
            Password
          </label>
        </div>
        
        <div
          className="row center-align"
        >
          <button
            className="waves-effect waves-light btn"
            type="submit"
          >
            Register
          </button>
        </div>
        <div
          className="row center-align"
        >
          Already have an account? 
          <a
            className="cursor-pointer blue-text"
            href="#!"
            onClick={
              [MockFunction] {
                "calls": Array [
                  Array [],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                ],
              }
            }
          >
            Login
          </a>
        </div>
      </div>
    </form>
  </div>
</ul>
`;

exports[`SideNav component should render component successfully 1`] = `
<ul
  className="sidenav"
  id="slide-out"
>
  <li
    className="git-logo"
  >
    <img
      alt=""
      className="logo-img"
      src="https://res.cloudinary.com/heavyman/image/upload/v1544835375/Git-Icon-Black_b2dvxp.png"
    />
    <p>
      The awesome git cheat sheet
    </p>
  </li>
  <li>
    <div
      className="divider"
    />
  </li>
  <li
    className="info-text"
  >
    <p>
      Whether you are new to git or just needing a refresher, this cheat sheet will help you discover or remember useful git commands.
    </p>
    <p
      className="center-align"
    >
      Click 
      <a
        className="cursor-pointer blue-text"
        href="#!"
        onClick={[Function]}
      >
        here 
      </a>
      for more privilege.
    </p>
  </li>
</ul>
`;

exports[`SideNav component should show login form when user clicks to login 1`] = `
<ul
  className="sidenav"
  id="slide-out"
>
  <li
    className="git-logo"
  >
    <img
      alt=""
      className="logo-img"
      src="https://res.cloudinary.com/heavyman/image/upload/v1544835375/Git-Icon-Black_b2dvxp.png"
    />
    <p>
      The awesome git cheat sheet
    </p>
  </li>
  <li>
    <div
      className="divider"
    />
  </li>
  <div
    className="row"
  >
    <form
      className="col s12"
      onSubmit={[Function]}
    >
      <div
        className="row"
      >
        <div
          className="input-field col s12"
        >
          <input
            className="validate"
            id="username"
            name="username"
            onChange={[Function]}
            required={true}
            type="text"
          />
          <label
            htmlFor="username"
          >
            Username
          </label>
        </div>
        <div
          className="input-field col s12"
        >
          <input
            className="validate"
            id="password"
            name="password"
            onChange={[Function]}
            required={true}
            type="password"
          />
          <label
            htmlFor="password"
          >
            Password
          </label>
        </div>
        
        <div
          className="row center-align"
        >
          <button
            className="waves-effect waves-light btn"
            type="submit"
          >
            Login
          </button>
        </div>
        <div
          className="row center-align"
        >
          Don't have an account? 
          <a
            className="cursor-pointer blue-text"
            href="#!"
            onClick={[Function]}
          >
            Register
          </a>
        </div>
      </div>
    </form>
  </div>
</ul>
`;
